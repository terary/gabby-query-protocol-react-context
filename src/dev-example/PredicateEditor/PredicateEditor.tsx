/* eslint-disable react/require-default-props */
/* eslint-disable import/prefer-default-export */
/* eslint-disable no-console */
import React, { useEffect, useState } from "react";
import type {
  TPredicateProperties,
  TPredicateOperator,
  TPredicatePropertiesArrayValue,
  TPredicateSubjectWithId,
  TGabbyQueryProtocolContextType,
} from "../../lib";

import InputGeneric from "./InputGeneric";
import SelectGeneric from "./SelectGeneric";
import SelectMultipleGeneric from "./SelectMultipleGeneric";
import { GabbyQueryProtocolContext } from "../../lib";

const noop = (...args: unknown[]) => {
  // console.log("PredicateBuilder noop called");
  // console.dir(args);
};

const emptyPredicate = (
  subject: TPredicateSubjectWithId
): TPredicateProperties | TPredicatePropertiesArrayValue => {
  const { subjectId, validOperators } = { ...subject };

  if (subjectId === undefined || validOperators === undefined) {
    return {} as TPredicateProperties;
  }

  const operator = Object.keys(validOperators).shift();

  return {
    value: "",
    subjectId,
    operator,
  } as TPredicateProperties;
};

interface PredicateEditorProps {
  // readonly querySubject: TQuerySubjectWithId;
  subjectId: string;
  initialPredicate?: TPredicateProperties | TPredicatePropertiesArrayValue;
  onFinish?: (predicate: TPredicateProperties) => void;
  onCancel?: () => void;
}
export const PredicateEditor = ({
  initialPredicate = {} as TPredicateProperties | TPredicatePropertiesArrayValue,
  subjectId,
  onFinish = noop,
  onCancel = noop,
}: PredicateEditorProps): JSX.Element => {
  const { subjectDictionary, operatorLabels } = React.useContext(
    GabbyQueryProtocolContext
  ) as TGabbyQueryProtocolContextType;

  const querySubject = subjectDictionary.getSubject(subjectId);

  const [currenTPredicateProperties, setCurrenTPredicateProperties] = useState({
    ...emptyPredicate(querySubject),
    ...initialPredicate,
  } as TPredicateProperties | TPredicatePropertiesArrayValue);

  const optionLists = subjectDictionary.getOptionsList(subjectId);

  useEffect(() => {
    setCurrenTPredicateProperties({
      ...emptyPredicate(querySubject),
      ...initialPredicate,
    });
  }, [subjectId]);

  const handleOperatorChange = (ev: React.ChangeEvent<HTMLSelectElement>): void => {
    const newPredicate = {
      ...currenTPredicateProperties,
      ...{ operator: ev.target.value as TPredicateOperator },
    };
    setCurrenTPredicateProperties(newPredicate);
  };

  const handleMultiValueChange = (newValue: (string | number)[]) => {
    setCurrenTPredicateProperties({
      ...currenTPredicateProperties,
      ...{ value: newValue },
    } as TPredicatePropertiesArrayValue);
  };
  const handleValueChange = (newValue: string | number) => {
    setCurrenTPredicateProperties({
      ...currenTPredicateProperties,
      ...{ value: newValue },
    });
  };

  const Operators = () => {
    return (
      <>
        <select
          value={currenTPredicateProperties.operator}
          onChange={handleOperatorChange}
        >
          {Object.keys(querySubject.validOperators).map((op) => {
            return (
              <option value={op} key={op}>
                {operatorLabels[op as TPredicateOperator]}({op})
              </option>
            );
          })}
        </select>
      </>
    );
  };

  const handleFinish = () => {
    onFinish(currenTPredicateProperties);
  };

  const isSelectOperator = () => {
    return (
      currenTPredicateProperties.operator === "$nanyOf" ||
      currenTPredicateProperties.operator === "$anyOf" ||
      currenTPredicateProperties.operator === "$oneOf"
    );
  };

  const isMultiSelect = () => {
    return ["$nanyOf", "$anyOf"].includes(currenTPredicateProperties.operator);
  };
  return (
    <>
      <Operators />
      {isSelectOperator() && isMultiSelect() && (
        <SelectMultipleGeneric
          options={
            optionLists[
              currenTPredicateProperties.operator as "$nanyOf" | "$anyOf" | "$oneOf"
            ]
          }
          value={currenTPredicateProperties.value}
          onChange={handleMultiValueChange}
        />
      )}
      {isSelectOperator() && !isMultiSelect() && (
        <SelectGeneric
          options={
            optionLists[
              currenTPredicateProperties.operator as "$nanyOf" | "$anyOf" | "$oneOf"
            ]
          }
          value={currenTPredicateProperties.value}
          onChange={handleValueChange}
        />
      )}

      {!isSelectOperator() && (
        <InputGeneric
          value={currenTPredicateProperties.value}
          onChange={handleValueChange}
        />
      )}

      <button onClick={handleFinish} type="button">
        <span style={{ color: "green" }}>&#10004;</span>
      </button>
      <button onClick={onCancel} type="button">
        <span style={{ color: "red" }}>&#10006;</span>
      </button>
    </>
  );
};
