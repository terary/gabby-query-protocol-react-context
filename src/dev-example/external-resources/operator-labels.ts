// cspell:disable
// import type { TPredicateOperatorLabels } from "gabby-query-protocol-lib";
import { TPredicateOperatorLabels } from "../../lib/GabbyQueryProtocolContext";

const EN: TPredicateOperatorLabels = {
  $and: "All of these",
  $anyOf: "Any Of",
  $empty: "Is Empty",
  $eq: "Equals",
  $gt: "Greater Than",
  $gte: "Greater Than Equal",
  $isnull: "Is Null",
  $lt: "Less Than",
  $lte: "Less Than Equal",
  $like: "Contains",
  $oneOf: "One Of",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "Any of these",
};
const ES: TPredicateOperatorLabels = {
  $and: "Todos estos",
  $anyOf: "Cualquiera de",
  $empty: "Is Empty",
  $eq: "Es igual a",
  $gt: "Mayor que",
  $gte: "Mayor que igual",
  $isnull: "Is Null",
  $lt: "Menor que",
  $lte: "Menor que igual",
  $like: "Contiene",
  $oneOf: "Uno de",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "Cualquiera de estos",
};

const TH: TPredicateOperatorLabels = {
  $and: "ทั้งหมดนี้",
  $anyOf: "อันใดอันหนึ่ง",
  $empty: "Is Empty",
  $eq: "เท่ากับ",
  $gt: "มากกว่า",
  $gte: "มากกว่าเท่ากับ",
  $isnull: "Is Null",
  $lt: "น้อยกว่า",
  $lte: "น้อยกว่าเท่ากับ",
  $like: "มี",
  $oneOf: "หนึ่งใน",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "ใดๆ เหล่านี้",
};
const AR: TPredicateOperatorLabels = {
  $and: "كل هذه",
  $anyOf: "اي من",
  $empty: "Is Empty",
  $eq: "يساوي",
  $gt: "أكبر من",
  $gte: "أكبر من يساوي",
  $isnull: "Is Null",
  $lt: "أقل من",
  $lte: "أقل من يساوي",
  $like: "يحتوي على",
  $oneOf: "واحد من",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",

  $or: "أي منهم",
};

const SymbolsTypable: TPredicateOperatorLabels = {
  $and: "AND",
  $anyOf: "Any Of",
  $empty: "Is Empty",
  $eq: "=",
  $gt: ">",
  $gte: ">=",
  $isnull: "Is Null",
  $lt: "<",
  $lte: "<=",
  $like: "=~",
  $oneOf: "One Of",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",

  $or: "OR",
};
const SymbolsEntities: TPredicateOperatorLabels = {
  // https://dev.w3.org/html5/html-author/charref
  $anyOf: String.fromCharCode(0x02286), // ⊆
  $empty: "Is Empty",

  $eq: "=",
  $gt: ">",
  $gte: String.fromCharCode(0x02265), // ≥
  $isnull: "Is Null",
  $lt: "<",
  $lte: String.fromCharCode(0x02264), // ≤
  $like: String.fromCharCode(0x02243), // ≃
  $oneOf: String.fromCharCode(0x02203), // ∃
  $and: String.fromCharCode(0x02227), // ∧
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: String.fromCharCode(0x02228), // ∨
};

export { SymbolsEntities, SymbolsTypable, AR, EN, ES, TH };

// probably a more eloquent way of doing this but ran out of time.
// export const PREDICATE_OPERATORS = ["$eq", "$gt", "$lt", "$gte", "$lte", "$like", "$in"];
// export const JUNCTION_OPERATORS = ["$and", "$or"];
// export const predefinedLabels = {
//   Symbols,
//   en: Long,
//   "en-US": Long,
//   es: ES,
//   "es-MX": ES,
//   "th-TH": TH,
//   AR,
// };
