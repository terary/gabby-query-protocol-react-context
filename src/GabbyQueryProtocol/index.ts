// ------------------ projection
import {
  CONSTS as PROJECTION_CONSTS,
  EXAMPLE_JSON_BLUE_SKIES as PROJECTION_EXAMPLE_JSON_BLUE_SKIES,
} from "gabby-query-protocol-projection";
import {
  ProjectionContext,
  useProjectionSubjects,
  useProjectionUtilities,
} from "./Projections";
import type { TProjectionContextType } from "./Projections";

// ------------------ Predicate Formula
import {
  CONSTS as LIB_CONSTS,
  EXAMPLE_JSON_BLUE_SKIES as LIB_EXAMPLE_JSON_BLUE_SKIES,
} from "gabby-query-protocol-lib";
import type { TPredicateFormulaEditorContextType } from "./PredicateFormula";

import {
  PredicateFormulaEditorContextHooks,
  PredicateFormulaEditorContext,
} from "./PredicateFormula";

// ------------------ Misc
import * as opLabelsI18N from "./external-resources/operator-labels";

const GABBY_EXAMPLE_JSON_BLUE_SKIES = {
  LIB: LIB_EXAMPLE_JSON_BLUE_SKIES,
  PROJECTION: PROJECTION_EXAMPLE_JSON_BLUE_SKIES,
};

const GABBY_CONSTS = {
  LIB: LIB_CONSTS,
  PROJECTION: PROJECTION_CONSTS,
};

// ------------------------ Export
export { GABBY_EXAMPLE_JSON_BLUE_SKIES, GABBY_CONSTS, opLabelsI18N };

export const ProjectionContextHooks = {
  useProjectionSubjects,
  useProjectionUtilities,
};

export {
  PredicateFormulaEditorContext,
  PredicateFormulaEditorContextHooks,
  ProjectionContext,
  // ProjectionContextHooks,
};

export type { TPredicateFormulaEditorContextType, TProjectionContextType };
