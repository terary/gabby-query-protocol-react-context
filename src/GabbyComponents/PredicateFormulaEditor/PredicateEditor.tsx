/* eslint-disable react/require-default-props */
/* eslint-disable import/prefer-default-export */

import React, { useState } from "react";
import InputLabel from "@mui/material/InputLabel";
import FormControl from "@mui/material/FormControl";
import Select, { SelectChangeEvent } from "@mui/material/Select";
import { useTranslation } from "react-i18next";
import { useTheme } from "@mui/system";
import { GridSize } from "@mui/material";
import Grid from "@mui/material/Grid";
import type {
  TPredicateOperator,
  TPredicateProperties,
  TPredicatePropertiesArrayValue,
  TPredicateSubjectProperties,
  TPredicateSubjectWithId,
  TValidatorResponse,
  TValidOperatorList,
} from "gabby-query-protocol-lib";
import { ButtonsFinishCancelText } from "../common/ButtonsFinishCancelText";
import { ButtonsFinishCancelIcon } from "../common/ButtonsFinishCancelIcon";
import { InputMux } from "./InputMux";
// import { usePredicateTreeUtilities } from "../../GabbyQueryProtocol/PredicateFormula";
import { PredicateFormulaEditorContextHooks } from "../../GabbyQueryProtocol";

import { Validators } from "gabby-query-protocol-lib";
import type {
  TPredicateNodeJson,
  IPredicateSubjectDictionary,
} from "gabby-query-protocol-lib";
const { usePredicateTreeUtilities } = PredicateFormulaEditorContextHooks;

export interface IValidatePredicateAgainstOperator {
  (
    predicateJson: TPredicateNodeJson,
    subjects: IPredicateSubjectDictionary
  ): TValidatorResponse;
}

// TODO - *tmc* - this should be exported/imported from package somewhere?
type ValueType = number | string | (number | string)[]; // do not export.  If export is necessary change name T... (no "Type"), the convention

const getFirstValidOperator = (validOperators: TValidOperatorList): TPredicateOperator => {
  return Object.keys(validOperators)[0] as TPredicateOperator;
};

const breakpoints = {
  "1row": {
    subject: 3 as GridSize,
    operator: 3 as GridSize,
    value: 5 as GridSize,
    controlButtons: 1 as GridSize,
  },
  "2row": {
    subject: 6 as GridSize,
    operator: 6 as GridSize,
    value: 10 as GridSize,
    controlButtons: 2 as GridSize,
  },
  "3row": {
    subject: 12 as GridSize,
    operator: 12 as GridSize,
    value: 12 as GridSize,
    controlButtons: 12 as GridSize,
  },
};

const noOpValidator = (
  predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
): TValidatorResponse => {
  return { hasError: false, errorMessages: [] };
};

const noopTestSpyPredicateOnChange = (
  predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
): void => {};

interface Props {
  layout?: "1row" | "2row" | "3row";
  initialPredicate: TPredicateProperties | TPredicatePropertiesArrayValue;
  onCancel: () => void;
  onFinish: (
    predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
  ) => void;
  testSpyPredicateOnChange?: (
    predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
  ) => void;

  validator?: IValidatePredicateAgainstOperator;
  // validator?: (
  //   predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
  // ) => TValidatorResponse;
}

export const PredicateEditor = ({
  layout = "1row",
  initialPredicate,
  onCancel,
  onFinish,
  testSpyPredicateOnChange = noopTestSpyPredicateOnChange,
  validator = Validators.ValidatePredicateAgainstOperator,
}: Props): JSX.Element => {
  const theme = useTheme();
  const { subjectDictionary, operatorLabels } = usePredicateTreeUtilities();

  const [userMessage, setUserMessage] = useState<string | null>(null);
  const [selectedSubject, setSelectedSubject] = React.useState<TPredicateSubjectWithId>(
    subjectDictionary.getSubject(initialPredicate.subjectId)
  );

  const [predicateExpression, setPredicateExpression] = React.useState<
    TPredicateProperties | TPredicatePropertiesArrayValue
  >(initialPredicate);

  const _updatePredicateState = (
    newPredicate: TPredicateProperties | TPredicatePropertiesArrayValue
  ) => {
    testSpyPredicateOnChange(newPredicate);
    setPredicateExpression(newPredicate);
  };

  const handleSelectedSubjectChange = (event: SelectChangeEvent) => {
    const subjectId = event.target.value as string;
    const subjectProps = subjectDictionary.getSubject(subjectId);

    const operator = getFirstValidOperator(subjectProps.validOperators);
    const newPredicate = { subjectId, operator, value: "" };

    _updatePredicateState(newPredicate);

    setSelectedSubject(subjectProps);
  };

  const handleOperatorChange = (event: SelectChangeEvent) => {
    const operator = event.target.value as TPredicateOperator;
    _updatePredicateState({ ...predicateExpression, operator, value: "" });
  };

  const handleValueChange = (value: ValueType) => {
    const pExpression = { ...predicateExpression };

    //@ts-ignore - there seems to be an issue with the way value is typed.
    // should be resolved in future releases
    pExpression.value = value;
    // setPredicateExpression(pExpression);
    _updatePredicateState(pExpression);
  };

  const handleFinishClick = () => {
    const subject = subjectDictionary.getSubject(
      predicateExpression.subjectId
    ) as TPredicateSubjectProperties;

    // TODO - *tmc* - this cast should not be required.  Issue is with validator
    if (subject.datatype === "integer") {
      predicateExpression.value = Number(predicateExpression.value);
    }
    const validatorResponse = validator(predicateExpression, subjectDictionary);
    if (validatorResponse.hasError) {
      console.log("Validation Debug:", validatorResponse.errorMessages);
      setUserMessage("Failed validation check console for more information");
    } else {
      setUserMessage(null);
      onFinish(predicateExpression);
    }
  };

  const SubjectSelectFormControl = () => {
    return (
      <FormControl variant="outlined" sx={{ width: "100%" }}>
        <InputLabel htmlFor="editor-select-subject">Subject</InputLabel>
        <Select
          inputProps={{
            name: "subjectId",
            id: "editor-select-subject",
            "data-testid": "test-subject-select",
          }}
          label="Subject"
          native
          onChange={handleSelectedSubjectChange}
          value={selectedSubject.subjectId}
        >
          {subjectDictionary.getSubjectIds().map((subjectId) => {
            const subject = subjectDictionary.getSubject(subjectId);
            return (
              <option key={subjectId} value={subjectId}>
                {subject.defaultLabel || subjectId}
              </option>
            );
          })}
        </Select>
      </FormControl>
    );
  };

  const OperatorSelectFormControl = () => {
    return (
      <FormControl variant="outlined" sx={{ width: "100%" }}>
        <InputLabel htmlFor="editor-select-operator">Operator</InputLabel>
        <Select
          native
          value={predicateExpression.operator}
          onChange={handleOperatorChange}
          label="Operator"
          inputProps={{
            name: "operator",
            id: "editor-select-operator",
          }}
        >
          {Object.entries(selectedSubject.validOperators).map(([op, operatorAttributes]) => {
            return (
              <option key={op} value={op}>
                {operatorLabels[op as TPredicateOperator]}
              </option>
            );
          })}
        </Select>
      </FormControl>
    );
  };

  return (
    <Grid container spacing={1}>
      <Grid item xs={12} md={breakpoints[layout].subject}>
        <SubjectSelectFormControl />
      </Grid>
      <Grid item xs={12} md={breakpoints[layout].operator}>
        <OperatorSelectFormControl />
      </Grid>
      <Grid item xs={12} md={breakpoints[layout].value} style={{ textAlign: "left" }}>
        <InputMux
          id="mux-id"
          name="mux-name"
          predicate={predicateExpression}
          onChange={handleValueChange}
        />
      </Grid>
      <Grid
        item
        xs={12}
        md={breakpoints[layout].controlButtons}
        style={{ textAlign: "center" }}
      >
        {layout === "1row" && (
          <ButtonsFinishCancelIcon onFinish={handleFinishClick} onCancel={onCancel} />
        )}
        {layout === "2row" && (
          <ButtonsFinishCancelIcon onFinish={handleFinishClick} onCancel={onCancel} />
        )}
        {layout === "3row" && (
          <ButtonsFinishCancelText onFinish={handleFinishClick} onCancel={onCancel} />
        )}
      </Grid>
      {userMessage !== null && (
        <Grid
          item
          xs={12}
          alignContent="center"
          textAlign="center"
          style={{
            color: theme.palette.error.main,
            marginLeft: "15px",
            marginRight: "15px",
          }}
        >
          {userMessage}
        </Grid>
      )}
    </Grid>
  );
};
