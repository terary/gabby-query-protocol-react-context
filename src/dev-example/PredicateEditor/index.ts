/* eslint-disable import/prefer-default-export */
import { PredicateEditor } from "./PredicateEditor";

import { subjectsDocument } from "./subject-document";

export { PredicateEditor };

const str = JSON.stringify(subjectsDocument);
const json = JSON.parse(str);
