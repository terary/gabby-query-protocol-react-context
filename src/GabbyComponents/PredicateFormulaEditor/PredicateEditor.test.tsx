import React from "react";
import renderer from "react-test-renderer";
import {
  render,
  screen,
  act,
  fireEvent,
  cleanup,
  within,
  getByTestId,
  getByLabelText,
  findByDisplayValue,
  getByText,
  findByText,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import selectEvent from "react-select-event";

import {
  PredicateFormulaEditorFactory,
  TPredicateProperties,
  TPredicatePropertiesArrayValue,
  TValidatorResponse,
} from "gabby-query-protocol-lib";

//import { AppContextProviders } from "../Application/AppContextProviders";
import GQPPredicateEditorContextProvider from "../../GabbyQueryProtocol/PredicateFormula/context";
//import GQPPredicateEditorContextProvider from "../../GabbyQueryProtocol/contexts/GQPPredicateFormula/context";
import { GABBY_EXAMPLE_JSON_BLUE_SKIES } from "../../GabbyQueryProtocol";
import * as predefinedOperatorLabels from "../../GabbyQueryProtocol/external-resources/operator-labels";

import { PredicateEditor } from "./PredicateEditor";
import { executionAsyncResource } from "async_hooks";

const subjectDictionaryJson =
  GABBY_EXAMPLE_JSON_BLUE_SKIES.LIB.predicateSubjectsDictionaryJson;
const predicateFormula = GABBY_EXAMPLE_JSON_BLUE_SKIES.LIB.predicateTreeJson;

const predicateFormulaEditor = PredicateFormulaEditorFactory.fromJson({
  subjectDictionaryJson,
  predicateTreeJson: predicateFormula,
});

type PredicateTypes = TPredicateProperties | TPredicatePropertiesArrayValue;

const validator = (predicateProperties: PredicateTypes) => {
  return { hasError: false, errorMessages: [] };
};

const noopFinish = (predicate: PredicateTypes) => {};
const noopCancel = () => {};
type Props = {
  onFinish?: (predicate: PredicateTypes) => void;
  onCancel?: () => void;
  testSpyPredicateOnChange?: (
    predicateProperties: TPredicateProperties | TPredicatePropertiesArrayValue
  ) => void;
};

const PredicateEditorWithApplicationContext = ({
  onFinish = noopFinish,
  onCancel = noopCancel,
  testSpyPredicateOnChange,
}: Props) => {
  return (
    <GQPPredicateEditorContextProvider
      predicateFormulaEditor={predicateFormulaEditor}
      operatorLabels={predefinedOperatorLabels.EN}
    >
      <PredicateEditor
        initialPredicate={predicateFormulaEditor.subjectDictionary.makeEmptyPredicate()}
        onFinish={(p: any) => {}}
        onCancel={() => {}}
        testSpyPredicateOnChange={testSpyPredicateOnChange}
        validator={validator}
      />{" "}
    </GQPPredicateEditorContextProvider>
  );
};
describe("PredicateEditor", () => {
  afterEach(cleanup);
  it("Should be awesome", () => {
    expect("Awesome").toBe("Awesome");
  });
  describe("SubjectChange", () => {
    it.only("Should change subject, reset operator and empty value", async () => {
      const changeObserver = jest.fn((p: any) => {
        console.log(p);
        renderer;
      });

      const mockFinish = jest.fn();

      const result = render(
        <PredicateEditorWithApplicationContext
          testSpyPredicateOnChange={changeObserver}
          onFinish={mockFinish}
        />
      );
    });
  });
  describe("Snapshots", () => {
    it.skip("renders correctly", () => {
      const tree = renderer.create(<PredicateEditorWithApplicationContext />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
