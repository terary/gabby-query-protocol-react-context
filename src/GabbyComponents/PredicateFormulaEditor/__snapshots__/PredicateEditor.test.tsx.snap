// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredicateEditor Snapshots renders correctly 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-1mhtw2e-MuiGrid-root"
    >
      <div
        className="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={true}
          htmlFor="editor-select-subject"
        >
          Subject
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <select
            aria-invalid={false}
            className="MuiNativeSelect-select MuiNativeSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="editor-select-subject"
            multiple={false}
            name="subjectId"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            value="firstName"
          >
            <option
              aria-label="None"
              value=""
            />
            <option
              value="firstName"
            >
              First Name
            </option>
            <option
              value="lastName"
            >
              Last Name
            </option>
            <option
              value="postalCode"
            >
              Postal Code
            </option>
            <option
              value="annualIncome"
            >
              Annual Income
            </option>
            <option
              value="region"
            >
              Region
            </option>
            <option
              value="favoriteFruit"
            >
              Favorite Fruit
            </option>
            <option
              value="startDate"
            >
              Start Date
            </option>
            <option
              value="numberOfChildren"
            >
              Number of Children
            </option>
          </select>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-186xcr5"
            >
              <span>
                Subject
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-1mhtw2e-MuiGrid-root"
    >
      <div
        className="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={true}
          htmlFor="editor-select-operator"
        >
          Operator
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <select
            aria-invalid={false}
            className="MuiNativeSelect-select MuiNativeSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="editor-select-operator"
            multiple={false}
            name="operator"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            value="$eq"
          >
            <option
              value="$eq"
            >
              يساوي
            </option>
            <option
              value="$gt"
            >
              أكبر من
            </option>
            <option
              value="$like"
            >
              يحتوي على
            </option>
          </select>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-186xcr5"
            >
              <span>
                Operator
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-5 css-1a8jlry-MuiGrid-root"
      style={
        Object {
          "textAlign": "left",
        }
      }
    >
      <div
        className="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="mux-id"
            id="mux-id-label"
          >
            Value
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id="mux-id"
              name="mux-name"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-dhh0px"
              >
                <span>
                  Value
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-1 css-mqqb9l-MuiGrid-root"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-uinpbw-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
        >
          <button
            aria-label="Finalizado"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeMedium css-1r6h6t4-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DoneSharpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
              />
            </svg>
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
        >
          <button
            aria-label="Cancelar"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-1jo1f9u-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ClearSharpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  " ",
]
`;
